### PR TITLE
EZP-24952: [VCL] Clarify User hash documentation

### PR DESCRIPTION
**Pull request by andrerom**

### CODE REVIEW

**Comment by joaoinacio** [comment]
+1
IMHO this resolves https://jira.ez.no/browse/EZP-24952

File: doc/varnish/vcl/varnish3.vcl [hunk 1]
```diff
@@ -138,8 +138,14 @@ sub ez_user_hash {
...
 
     if (req.restarts == 0 && (req.request == "GET" || req.request == "HEAD")) {
-        // Anonymous user => Set a hardcoded anonymous hash
+        // Get User (Context) hash, for varying cache by what user has access to.
+        // https://doc.ez.no/display/EZP/Context+aware+HTTP+cache
+
+        // Anonymous user => Set a hardcoded anonymous hash to avoid backend lookup for hash
         if (req.http.Cookie !~ "eZSESSID" && !req.http.authorization) {
+            // You may update this hash with the actual one for anonymous user
+            // to archive a bit better cache hit ratio across anonymous users.
```
Comment (lolautruche) [inline doc/varnish/vcl/varnish3.vcl]: s/archive/achieve ? Still not sure about the word though. Maybe just _to get a better cache hit ratio_?

File: doc/varnish/vcl/varnish4.vcl [hunk 2]
```diff
@@ -135,8 +135,14 @@ sub ez_user_hash {
...
 
     if (req.restarts == 0 && (req.method == "GET" || req.method == "HEAD")) {
-        // Anonymous user => Set a hardcoded anonymous hash
+        // Get User (Context) hash, for varying cache by what user has access to.
+        // https://doc.ez.no/display/EZP/Context+aware+HTTP+cache
+
+        // Anonymous user => Set a hardcoded anonymous hash to avoid backend lookup for hash
         if (req.http.Cookie !~ "eZSESSID" && !req.http.authorization) {
+            // You may update this hash with the actual one for anonymous user
+            // to archive a bit better cache hit ratio across anonymous users.
```
Comment (lolautruche) [inline doc/varnish/vcl/varnish4.vcl]: Same as above

**Comment by lolautruche** [comment]
So, as far as I understand this, suggestion would ensure that we don't have 2 identical variations of cache of anonymous users (with and without session)?

If so, +1

**Comment by andrerom** [comment]
> So, as far as I understand this, suggestion would ensure that we don't have 2 identical variations of cache of anonymous users (with and without session)?

yes

File: doc/varnish/vcl/varnish4.vcl [hunk 3]
```diff
@@ -135,8 +135,14 @@ sub ez_user_hash {
...
     if (req.restarts == 0 && (req.method == "GET" || req.method == "HEAD")) {
-        // Anonymous user => Set a hardcoded anonymous hash
+        // Get User (Context) hash, for varying cache by what user has access to.
+        // https://doc.ez.no/display/EZP/Context+aware+HTTP+cache
+
+        // Anonymous user w/o session => Use hardcoded anonymous hash to avoid backend lookup for hash
         if (req.http.Cookie !~ "eZSESSID" && !req.http.authorization) {
+            // You may update this hash with the actual one for anonymous user
+            // to get a better cache hit ratio across anonymous users.
+            // Note: You should then update it every time anonymous user rights change.
```
Comment (joaoinacio) [inline doc/varnish/vcl/varnish4.vcl]: Nitpick: 
`every time anonymous user rights change.` - note that modifying the anonymous role (adding/removing policies) has no effect in the hash AFAIK, only changing role(s) and assignment(s) for the user/group does.

Comment (andrerom) [inline doc/varnish/vcl/varnish4.vcl]: @lolautruche do you remember this part to verify?
 is it role id's + role assignment limitations? Or does it also loop over policies?

If @joaoinacio is right, then maybe:
`// Note: You should then update it every time Anonymous User inherited Role Assignments change.`

Comment (andrerom) [inline doc/varnish/vcl/varnish4.vcl]: no feedback, feel free to propose change afterwards, but sentence as-is is fine, it simply does not specify this.

**Pull request merged**